### ISSUE TITLE
quotes?

### ISSUE THREAD

**Issue by gurgeous** [ruby-formatter/rufo#74]
One of my favorite features from `rubocop --auto-correct` and `prettier` is quote cleanup. They will happily turn all single quotes into double quotes (or vice versa depending on config). Is this something rufo can do? That would be a great feature.

This is a bit like the tabs vs. spaces debate, and having rufo enforce the rule would obviate the need for discussion.

**Comment by gingermusketeer** [comment]
@gurgeous This is definitely something that rufo should be doing I think. There is currently a new branch `new-formatter` that is quite a significant change so unsure where to recommend that you send a PR...

It would be great to get some thoughts on what you think the behaviour should be. For example:
- Does it need to be configurable? (Ideally rufo won't be too configurable)
  - If it should be configurable what the configuration options should be.
- When should we be using single vs double quotes.


Once this is worked out then a PR with failing tests would be a great starting point! Something like this perhaps:

```
#~# ORIGINAL
#~# PENDING
#~# quote_style: :prefer_single

"a great string"

#~# EXPECTED

'a great string'

#~# ORIGINAL
#~# PENDING
#~# quote_style: :prefer_double

'a great string'

#~# EXPECTED

"a great string"

```

**Comment by gurgeous** [comment]
Great! In terms of single vs. double, here are a few data points:

- negligible performance difference (https://www.viget.com/articles/just-use-double-quoted-ruby-strings/)
- most of the JS world has settled on single quotes
- rubocop picked single quotes, lots of drama
- prettier defaults to double quotes, but only for legacy reasons (https://github.com/prettier/prettier/issues/1105)

For a long time I used double quotes, but my most recent company uses single quotes to make it easier to switch between Ruby and Javascript. Personally I think being decisive is more important than which one is picked. I have no preference, which is why I want to use a formatter :)

Perhaps we should make single quotes the default?

Prettier has some thoughtful ideas about quoting that we should emulate. From https://github.com/prettier/prettier/blob/master/docs/rationale.md#strings:

> Prettier enforces double quotes by default, but has a setting for enforcing single quotes instead. There are two exceptions:
>
> The number of escaped quotes are minimized. For example, if you have a string with a single quote inside, it will be enclosed in double quotes regardless of the quote setting: "that's a double quote", not 'that\'s a double quote'.
> ...
> (removed second exception which refers to JSX)
>
> Prettier maintains the way your string is escaped. For example, "🙂" won't be formatted into "\uD83D\uDE42" and vice versa.

I'll start by creating some failing tests.

**Comment by gurgeous** [comment]
This is sorta working in a branch. What do you think of this approach? The new quote tests will pass, though it obviously breaks a few others...

```ruby
  # for quote conversion, only consider the simplest of strings
  def simple_string(node)
    inner = node[1][1..-1]
    return false if inner.length > 1
    inner = inner[0]
    return "" if !inner
    return false if inner[0] != :@tstring_content
    string = inner[1]
    string
  end

  def convert_quotes(node)
    # do we have a quote mismatch?
    correct = (quote_style == :single) ? "'" : '"'
    wrong   = (quote_style == :single) ? '"' : "'"
    return false if current_token_value != wrong

    # is it a simple string node?
    string = simple_string(node)
    return false if !string
    # don't allow slashes
    return false if string.include?("\\")
    # don't allow strings that include the other quote type
    return false if string.include?(correct)

    # success!
    write correct
    next_token
    consume_token :on_tstring_content if string != ""
    write correct
    next_token

    true
  end
```

**Comment by gingermusketeer** [comment]
Thanks for all the info :)

I personally like each quote style for different reasons.  Single quotes are nicer to type but double don't need to be changed for interpolating and make it easier to copy paste JSON etc.

I wonder if we can check which is the most common style on open source ruby projects. Be handy to know if add a lot of noise if a tool like this was added to a project.

The code sample looks pretty reasonable to me. Couple of things:
- Should the logic around deciding to change the string be a separate method?
- Rather than converting quotes I think it would be good to format strings
   - Always do the same thing regardless of what the quote style is or should be as apposed to just converting bad quote styles.

**Comment by gurgeous** [comment]
Great, thanks for the feedback. I wrote a quick script to calculate single vs. double for some major open source projects. I used something like this to count strings - `rg --no-filename -c -g '*.rb' "'[^']*'"`. That regex isn't perfect but it ought to do for our purposes. Here are the results for some major gems:

```
# mostly double
capistrano   - 98% double (30, 1662)
jekyll       - 96% double (182, 4684)
redis-rb     - 95% double (129, 2372)
rails        - 92% double (4940, 59652)
puma         - 71% double (489, 1173)
haml         - 59% double (962, 1372)

# pretty even
ruby-pg      - 54% double (1082, 1274)
httparty     - 53% double (918, 1046)
pry          - 53% double (2539, 2816)

# mostly single
devise       - 62% single (1865, 1165)
sidekiq      - 69% single (1556, 686)
faraday      - 72% single (823, 324)
sinatra      - 73% single (4854, 1808)
faker        - 74% single (1108, 396)
hashie       - 94% single (1649, 103)
```

A bit weighted toward double, though I think we'd be fine picking either default. Tell me your preference! :)

I'll put together a proper PR soon, perhaps today.

**Comment by gurgeous** [comment]
Can you take a look at this? It uses `:single` as the default, which will necessitate fixing some existing tests. Once we decide on the default I'll put together a proper PR. I have no preference...

https://github.com/gurgeous/rufo/commit/e3f12bfea3c381d1dcd735c6dd4a0f2d19125db9

**Comment by gingermusketeer** [comment]
Looks like a pretty simple change :) Perhaps `requote_string(node)` should be `format_simple_string(node)`.

@mjago @splattael @bessey @danielma What are your thoughts on this?

**Comment by splattael** [comment]
@gingermusketeer @gurgeous I like it :+1: 

Personally, I'd prefer `:double` to be the default. But that's me :wink:

**Comment by gurgeous** [comment]
Fine with me! I'll put a PR together...

**Comment by gurgeous** [comment]
bump

**Comment by BrianHawley** [comment]
By the way: If `:double` is the default (as it is at the time of the writing of this comment), it will make it easier for people to write code that will be easily convertible to Crystal. Crystal's strings are double-quoted, and single-quoted literals are characters. Consider this a vote in favor of `:double`.

**Comment by gingermusketeer** [comment]
@splattael Any update on the release for this?

**Comment by gingermusketeer** [comment]
This has been released as part of the 0.3.0 release. Thanks for the PR!
